packages/material-ui-lab/src/Masonry/Masonry.js

```diff
@@ -81,48 +106,84 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
-    const handleScroll = () => {
-      if (masonryRef.current.clientHeight === 1998 && !didWarn) {
-        console.warn(
-          [
-            'Material-UI: The Masonry can have the maximum height of 2,000px on Chrome browser.',
-            'Items that go beyond this height fail to be rendered on Chrome browser.',
-            'You can find more in this open issue: https://github.com/mui-org/material-ui/issues/27934',
-          ].join('\n'),
-        );
-
-        // eslint-disable-next-line react-hooks/exhaustive-deps
-        didWarn = true;
+    const handleResize = () => {
+      const columnHeights = [];
+      const numOfRows = [];
+      let skip = false;
+      Array.from(masonryRef.current.children).forEach((child) => {
+        if (child.className.includes('line-break') || skip) {
+          return;
+        }
+        const computedStyle = window.getComputedStyle(child);
+        // if any one of children is not rendered yet, container's height shouldn't be set;
+        // this is especially crucial for image masonry
+        if (computedStyle.height === '0px') {
+          skip = true;
+          return;
+        }
+        const order = computedStyle.order;
+        const height = Number(computedStyle.height.replace('px', ''));
+        columnHeights[order - 1] = columnHeights[order - 1]
+          ? columnHeights[order - 1] + Math.ceil(height)
+          : Math.ceil(height);
+        numOfRows[order - 1] = numOfRows[order - 1] ? numOfRows[order - 1] + 1 : 1;
+      });
+      if (!skip) {
+        setMaxColumnHeight(Math.max(...columnHeights));
+        setMaxNumOfRows(Math.max(...numOfRows));
       }
     };
+    let resizeObserver;
+    try {
+      resizeObserver = new ResizeObserver(handleResize);
+    } catch (err) {
+      resizeObserver = MockResizeObserver();
+    }
     const container = masonryRef.current;
-    container.addEventListener('scroll', handleScroll);
+    resizeObserver.observe(container);
     return () => {
-      container.removeEventListener('scroll', handleScroll);
+      resizeObserver.unobserve(container);
     };
   }, []);
 
   const handleRef = useForkRef(ref, masonryRef);
+
   return (
-    <MasonryContext.Provider value={contextValue}>
-      <MasonryRoot
-        as={component}
-        className={clsx(classes.root, className)}
-        ref={handleRef}
-        ownerState={ownerState}
-        {...other}
-      >
-        {children}
-      </MasonryRoot>
-    </MasonryContext.Provider>
+    <MasonryRoot
+      as={component}
+      className={clsx(classes.root, className)}
+      ref={handleRef}
+      ownerState={ownerState}
+      {...other}
+    >
+      {children}
+      {new Array(
+        columns - 1 ||
+          columns.xl - 1 ||
+          columns.lg - 1 ||
+          columns.md - 1 ||
+          columns.sm - 1 ||
+          columns.xs - 1,
+      )
+        .fill('')
+        .map((_, index) => (
+          <span key={index} className="line-break" style={lineBreakStyle} />
```

Comment:
I don't think we ever hard-code a classname.

Reply:
I replaced it with `classes.break`: https://github.com/mui-org/material-ui/pull/28059/commits/ecb00232564767c05528060fa2135af22ae85caf.